core/src/main/scala/com/softwaremill/react/kafka_api_opt1/Consumer.scala

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (C) 2014 - 2016 Softwaremill <http://softwaremill.com>
+ * Copyright (C) 2016 Lightbend Inc. <http://www.lightbend.com>
+ */
+package com.softwaremill.react.kafka_api_opt1
+
+import scala.collection.immutable
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+
+import akka.NotUsed
+import akka.stream.Graph
+import akka.stream.Inlet
+import akka.stream.Outlet
+import akka.stream.Shape
+import akka.stream.SourceShape
+import akka.stream.scaladsl.Flow
+import akka.stream.scaladsl.GraphDSL
+import akka.stream.scaladsl.Source
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.clients.consumer.KafkaConsumer
+import org.apache.kafka.clients.consumer.OffsetAndMetadata
+import org.apache.kafka.common.TopicPartition
+
+object Consumer {
+
+  // FIXME those should probably be real case classes (or value classes)?
+  type CommitMsg = Map[TopicPartition, OffsetAndMetadata]
+  type CommitConfirmation = Future[CommitMsg]
+
+  case class ConsumerShape[K, V](
+      commit: Inlet[CommitMsg],
+      messages: Outlet[ConsumerRecord[K, V]],
+      confirmation: Outlet[CommitConfirmation]
```

Comment:
> What about offering these semantic packages separately, in optimized form? We could offer a Source that encapsulates the immediate commit logic and produces elements at-most-once, and we could offer a Flow that produces messages and consumes commits. 

@rkuhn what is wrong with the simple one-liners me and @patriknw showed, using the simple `commit()` API (Option 3) returning a `Future[Done]`:

```scala
// At-most-once
source.mapAsync { msg => msg.transaction.commit().map(_ => msg) }.via(launchRocketLogic)

// At-least-once
source.mapAsync { msg => bank.creditMyAccout(msg.record.value.money).map(_ => msg.transaction.commit()) }
```

The issue with Option 1 style processing will be that the only thing that really composes with them is `map()`. Basically everything else like `filter()`, `mapConcat()`, any fan-in/-out will need custom stages and/or graphs to work with that option. I actually recommend to try out some processing patterns in code using both styles.